sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -858,6 +859,26 @@ object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
 }
 
 /**
+  * Push [[Limit]] operators through [[Join]] operators, iff the join type is outer joins.
+  * Adding extra [[Limit]] operators on top of the outer-side child/children.
+  */
+object PushLimitThroughOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case f @ Limit(expr, Join(left, right, joinType, joinCondition)) =>
+      joinType match {
+        case RightOuter =>
+          Limit(expr, Join(left, Limit(expr, right), joinType, joinCondition))
+        case LeftOuter =>
+          Limit(expr, Join(Limit(expr, left), right, joinType, joinCondition))
+        case FullOuter =>
+          Limit(expr, Join(Limit(expr, left), Limit(expr, right), joinType, joinCondition))
+        case _ => f // DO Nothing for the other join types
+      }
+  }
+}
```

Comment:
You are right, but, after this rule, there exists another rule `CombineLimits`, which can combine the extra limits. I think we should still fix it anyway. Thanks!